File: haskell/day_1/app/Main.hs
module Main where

import Data.List (tails)
import System.Environment (getArgs)

main :: IO ()
main = do
  args <- getArgs
  contents <- readFile $ head args

  let input = (\x -> read x :: Integer) <$> lines contents

  print $ "Part 1 answer: " ++ show (partOne input)
  print $ "Part 2 answer: " ++ show (partOne . map (\(a, b, c) -> a + b + c) $ window 3 input)

  return ()

window :: Int -> [a] -> [(a, a, a)]
window n l = map (\[a, b, c] -> (a, b, c)) . filter f . map (take n) $ tails l
  where
    f [_, _, _] = True
    f _ = False

partOne :: [Integer] -> Integer
partOne = snd . foldl f (Nothing, 0)
  where
    f (Nothing, acc) item = (Just item, acc)
    f (Just prev, acc) item = (Just item, if prev < item then acc + 1 else acc)

File: rust/day_1/src/main.rs
use std::env;
use std::fs;

fn main() {
    let mut args: Vec<String> = env::args().collect();

    // Trash command args
    args.remove(0);

    if let Some(filename) = args.pop() {
        let contents = fs::read_to_string(filename).expect("unable to read file");
        let input: Vec<u32> = contents
            .lines()
            .map(|l| l.parse::<u32>().expect("unable to parse input"))
            .collect();

        println!("Part 1 answer: {}", solve_part_1(&input));

        let input = input
            .iter()
            .zip(input.iter().skip(1))
            .zip(input.iter().skip(2))
            .map(|((a, b), c)| a + b + c)
            .collect::<Vec<u32>>();

        println!("Part 2 answer {}", solve_part_1(&input));
    } else {
        println!("No input file given");
    }
}

// Returns the number of times input increases from the previous value
fn solve_part_1(input: &Vec<u32>) -> u32 {
    let (_, acc) = input.into_iter().fold((None, 0), |(previous, acc), item| {
        let acc = match previous {
            Some(prev) if prev < item => acc + 1,
            _ => acc,
        };

        (Some(item), acc)
    });

    acc
}
